File: cses/04_graph_algorithms/1682_flight-routes-check/solution.py
import sys
from collections import deque


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1

    # Xây danh sách kề cho đồ thị gốc và đồ thị đảo chiều mọi cạnh
    adj = [[] for _ in range(n + 1)]   # đồ thị gốc
    radj = [[] for _ in range(n + 1)]  # đồ thị đảo chiều
    for _ in range(m):
        a = int(data[idx]); idx += 1
        b = int(data[idx]); idx += 1
        adj[a].append(b)
        radj[b].append(a)

    # BFS trả về mảng đánh dấu và số đỉnh tới được từ start trên đồ thị graph
    def bfs(graph, start):
        visited = bytearray(n + 1)
        visited[start] = 1
        count = 1
        q = deque([start])
        while q:
            u = q.popleft()
            for v in graph[u]:
                if not visited[v]:
                    visited[v] = 1
                    count += 1
                    q.append(v)
        return visited, count

    out = []

    # (1) Từ đỉnh 1 có tới được mọi đỉnh trong đồ thị gốc không?
    vis1, cnt1 = bfs(adj, 1)
    if cnt1 < n:
        for u in range(1, n + 1):
            if not vis1[u]:  # không đi được từ 1 tới u -> phản ví dụ (1, u)
                out.append("NO")
                out.append("%d %d" % (1, u))
                sys.stdout.write("\n".join(out) + "\n")
                return

    # (2) Từ đỉnh 1 có tới được mọi đỉnh trong đồ thị đảo chiều không?
    #     Tương đương: mọi đỉnh đều tới được đỉnh 1 trong đồ thị gốc.
    vis2, cnt2 = bfs(radj, 1)
    if cnt2 < n:
        for u in range(1, n + 1):
            if not vis2[u]:  # u không tới được đỉnh 1 -> phản ví dụ (u, 1)
                out.append("NO")
                out.append("%d %d" % (u, 1))
                sys.stdout.write("\n".join(out) + "\n")
                return

    # Cả hai lần BFS đều phủ hết n đỉnh -> đồ thị liên thông mạnh
    sys.stdout.write("YES\n")


main()

File: cses/04_graph_algorithms/1682_flight-routes-check/solution.cpp
#include <iostream>
#include <queue>
#include <vector>

// BFS trả về mảng đánh dấu những đỉnh tới được từ đỉnh start trên đồ thị graph
std::vector<char> reachable_from(
    const std::vector<std::vector<int>>& graph,
    const int start
) {
    std::vector<char> visited(graph.size(), false);
    std::queue<int> pending;
    visited[start] = true;
    pending.push(start);

    while (!pending.empty()) {
        const int city = pending.front();
        pending.pop();
        for (const int next : graph[city]) {
            if (!visited[next]) {
                visited[next] = true;
                pending.push(next);
            }
        }
    }
    return visited;
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;

    // Xây danh sách kề cho đồ thị gốc và đồ thị đảo chiều mọi cạnh
    std::vector<std::vector<int>> adj(n + 1);
    std::vector<std::vector<int>> reverse_adj(n + 1);
    for (int i = 0; i < m; ++i) {
        int a, b;
        std::cin >> a >> b;
        adj[a].push_back(b);
        reverse_adj[b].push_back(a);
    }

    // (1) Từ đỉnh 1 phải tới được mọi đỉnh trong đồ thị gốc,
    //     nếu không thì cặp (1, city) là phản ví dụ
    const std::vector<char> forward = reachable_from(adj, 1);
    for (int city = 1; city <= n; ++city) {
        if (!forward[city]) {
            std::cout << "NO\n1 " << city << '\n';
            return 0;
        }
    }

    // (2) Từ đỉnh 1 phải tới được mọi đỉnh trong đồ thị đảo chiều;
    //     tới được city ở đồ thị đảo nghĩa là city tới được 1 ở đồ thị gốc,
    //     nếu không thì cặp (city, 1) là phản ví dụ
    const std::vector<char> backward = reachable_from(reverse_adj, 1);
    for (int city = 1; city <= n; ++city) {
        if (!backward[city]) {
            std::cout << "NO\n" << city << " 1\n";
            return 0;
        }
    }

    // Cả hai lần BFS đều phủ hết n đỉnh -> đồ thị liên thông mạnh
    std::cout << "YES\n";
    return 0;
}
